src/main/kotlin/xyz/avalonxr/command/WelcomeCommand.kt

```diff
@@ -0,0 +1,160 @@
+package xyz.avalonxr.command
+
+import discord4j.common.util.Snowflake
+import discord4j.core.event.domain.interaction.ChatInputInteractionEvent
+import discord4j.core.`object`.entity.channel.TextChannel
+import org.springframework.beans.factory.annotation.Autowired
+import xyz.avalonxr.annotations.AvalonCommand
+import xyz.avalonxr.data.CommandResult
+import xyz.avalonxr.data.OptionStore
+import xyz.avalonxr.data.command.CommandBinding
+import xyz.avalonxr.data.entity.WelcomeMessage
+import xyz.avalonxr.data.error.AvalonError
+import xyz.avalonxr.data.error.CommandError
+import xyz.avalonxr.dsl.discord.OptionType
+import xyz.avalonxr.dsl.discord.command
+import xyz.avalonxr.dsl.discord.option
+import xyz.avalonxr.dsl.discord.optional
+import xyz.avalonxr.dsl.discord.subCommand
+import xyz.avalonxr.enums.CommandScope
+import xyz.avalonxr.models.discord.Command
+import xyz.avalonxr.repository.WelcomeMessageRepository
+import kotlin.jvm.optionals.getOrDefault
+import kotlin.jvm.optionals.getOrNull
+
+@AvalonCommand
+class WelcomeCommand @Autowired constructor(
+    private val messageRepository: WelcomeMessageRepository,
+) : Command {
+
+    override fun getBinding(): CommandBinding = CommandBinding(
+        command("welcome") {
+            description("Welcome command")
+
+            subCommand {
+                name("set")
+                description("Configure welcome message")
+
+                option(OptionType.STRING) {
+                    name("message")
+                    description("Welcome Message")
+                }
+
+                option(OptionType.CHANNEL) {
+                    name("channel")
+                    description("Welcome Channel")
+                    optional()
+                }
+            }
+
+            subCommand {
+                name("preview")
+                description("Preview welcome message")
+
+                option(OptionType.BOOLEAN) {
+                    name("formatted")
+                    description("Preview formatted or un-formatted message")
+                    optional()
+                }
+            }
+        },
+        CommandScope.GUILD
+    )
+
+    override fun processCommand(
+        source: ChatInputInteractionEvent,
+        options: OptionStore
+    ): CommandResult {
+        return when (options.subCommand) {
+            "set" -> setSubcommand(source, options)
+            "preview" -> previewSubcommand(source, options)
+            else -> welcomeMessageError(CommandError.CommandNotFound("[set, preview]"), source)
```

Comment:
It might be better to provide a sub-command specific error message for cases like this.